### PR TITLE
Fixed issues #38, #44 and #45

### PR DESCRIPTION
**Pull request by jonasscheid**
This PR addresses the following issues:
- #44
New NetMHCII 2.3 interface inherits from NetMHCII 2.2. The output between both predictors differed slightly, therefore methods like `parse_external_result` have been overwritten. Supported alleles have been extracted from the command line tool
New NetMHCIIpan 4.0 inherits from NetMHCIIpan 3.1. Again output differed slightly therefore `parse_external_result` has been overwritten. Supported alleles have been extracted from the command line tool.

- #45
As described in the issue  all class 2 predictors in `External.py` did not predict `CombinedAllele`s if they are on the list of supporting alleles due to a false internal representation of supported alleles. Changed all combined alleles in `__supported_alleles` from e.g.`HLA-DPA1*03:02-HLA-DPB1*58:01` to `HLA-DPA1*03:02-DPB1*58:01`

- #38
As described the input `Allele Obj` is converted to a string representation to plug it into a predictor and afterwards reverted into an `Allele Obj`. Therefore initial information (e.g. prob) is lost. This is also the case for the input `Peptide Obj`s. Addressed this issue in `ANN.py` by creating a dictionary `{Allele Obj: allele_str_representation}` in the `predict` method of each predictor.
In `External.py` this has been done with a dictionary `{allele_str_representation: Allele Obj}` since that was already present in the code. This has the consequence, that in each `parse_external_result` the input allele representation is not instantly converted to an `Allele Obj`. Instead this is done in the `predict` method using the dictionary.

### CODE REVIEW

File: epytope/EpitopePrediction/External.py [hunk 1]
```diff
@@ -168,14 +170,15 @@ def predict(self, peptides, alleles=None, command=None, options=None, **kwargs):
...
+                    for allele_string, scores in res_tmp.items():
+                        allele = alleles_string[allele_string]
                         if allele not in result.keys():
                             result[allele] = {}
                         for scoretype, pep_scores in scores.items():
                             if scoretype not in result[allele].keys():
                                 result[allele][scoretype] = {}
                             for pep, score in pep_scores.items():
-                                result[allele][scoretype][pep] = score
+                                result[allele][scoretype][pep_seqs[pep]] = score
```
Comment (christopher-mohr) [inline epytope/EpitopePrediction/External.py]: Why did this change? Seems unrelated...

File: epytope/EpitopePrediction/External.py [hunk 2]
```diff
@@ -473,12 +476,12 @@ def parse_external_result(self, file):
         scores = defaultdict(defaultdict)
         ranks = defaultdict(defaultdict)
         f = csv.reader(open(file, "r"), delimiter='\t')
-        alleles = [Allele(x.split()[0][:5]+'*'+x.split()[0][5:7]+ ':' +x.split()[0][7:]) for x in [x for x in next(f) if x.strip() != ""]]
+        alleles = [x.split()[0] for x in [x for x in next(f) if x.strip() != ""]]
```
Comment (christopher-mohr) [inline epytope/EpitopePrediction/External.py]: Same here and the next changes.

File: epytope/EpitopePrediction/External.py [hunk 3]
```diff
@@ -6538,9 +7864,12 @@ def parse_external_result(self, file):
                 if row[0] in ["#", "-"] or row.strip() == "" or "pos" in row:
                     continue
                 else:
-                    allele, pep, score = Allele(row.split()[HLAIndex.PICKPOCKET_1_1]), Peptide(row.split()[PeptideIndex.PICKPOCKET_1_1]), float(row.split()[ScoreIndex.PICKPOCKET_1_1])
+                    allele = row.split()[HLAIndex.PICKPOCKET_1_1].replace('*','')
```
Comment (christopher-mohr) [inline epytope/EpitopePrediction/External.py]: Why is the replacement of `*` necessary now?

**Review by christopher-mohr — CHANGES_REQUESTED** [review]

File: epytope/EpitopePrediction/External.py (hunk 3 above)
Comment (jonasscheid) [reply epytope/EpitopePrediction/External.py]: Yeah this is an ugly one by PickPocket.. That is because the allele input representation (`HLA-A0101`) differs from the allele output representation (`HLA-A*0101`). Since I save the allele input representation with the `Allele Obj` in a dictionary in `predict`, it needs to be changed to the allele input representation

File: epytope/EpitopePrediction/External.py (hunk 1 above)
Comment (jonasscheid) [reply epytope/EpitopePrediction/External.py]: I observed the same issue with newly created `Allele Obj` also with `Peptide Obj`. Therefore I used this `pep_seqs` dictionary to save the information of the `Peptide Obj(s)`

File: epytope/EpitopePrediction/External.py (hunk 2 above)
Comment (jonasscheid) [reply epytope/EpitopePrediction/External.py]: See changed description in PR

File: epytope/EpitopePrediction/External.py [hunk 4]
```diff
@@ -1454,9 +1457,9 @@ def parse_external_result(self, file):
         f = csv.reader(open(file, "r"), delimiter = '\t')
         scores = defaultdict(defaultdict)
         ranks = defaultdict(defaultdict)
-        alleles = [Allele(x[:5]+'*'+x[5:]) for x in next(f) if x != ""]
+        alleles = [x for x in next(f) if x != ""]
```
Comment (christopher-mohr) [inline epytope/EpitopePrediction/External.py]: I think it would be good if you would add information about the result dictionary to the `parse_external_result` method signature. Just to make sure that it is clear that we have `allele strings` etc.

File: epytope/EpitopePrediction/External.py [hunk 5]
```diff
@@ -2087,10 +2089,10 @@ def parse_external_result(self, file):
             if not len(row):
                 continue
             
-            if Allele.organism not in row[HLAIndex.NETMHCII_2_2]:
+            if "HLA" not in row[HLAIndex.NETMHCII_2_2]:
```
Comment (christopher-mohr) [inline epytope/EpitopePrediction/External.py]: I assume this did not work?

File: epytope/EpitopePrediction/External.py [hunk 6]
```diff
@@ -5853,6 +5985,1200 @@ def parse_external_result(self, file):
...
+'HLA-DQA1*06:02-DQB1*06:07', 'HLA-DQA1*06:02-DQB1*06:08', 'HLA-DQA1*06:02-DQB1*06:09', 'HLA-DQA1*06:02-DQB1*06:10', 'HLA-DQA1*06:02-DQB1*06:11',
+'HLA-DQA1*06:02-DQB1*06:12', 'HLA-DQA1*06:02-DQB1*06:14', 'HLA-DQA1*06:02-DQB1*06:15', 'HLA-DQA1*06:02-DQB1*06:16', 'HLA-DQA1*06:02-DQB1*06:17',
+'HLA-DQA1*06:02-DQB1*06:18', 'HLA-DQA1*06:02-DQB1*06:19', 'HLA-DQA1*06:02-DQB1*06:21', 'HLA-DQA1*06:02-DQB1*06:22', 'HLA-DQA1*06:02-DQB1*06:23',
+'HLA-DQA1*06:02-DQB1*06:24', 'HLA-DQA1*06:02-DQB1*06:25', 'HLA-DQA1*06:02-DQB1*06:27', 'HLA-DQA1*06:02-DQB1*06:28', 'HLA-DQA1*06:02-DQB1*06:29',
+'HLA-DQA1*06:02-DQB1*06:30', 'HLA-DQA1*06:02-DQB1*06:31', 'HLA-DQA1*06:02-DQB1*06:32', 'HLA-DQA1*06:02-DQB1*06:33', 'HLA-DQA1*06:02-DQB1*06:34',
+'HLA-DQA1*06:02-DQB1*06:35', 'HLA-DQA1*06:02-DQB1*06:36', 'HLA-DQA1*06:02-DQB1*06:37', 'HLA-DQA1*06:02-DQB1*06:38', 'HLA-DQA1*06:02-DQB1*06:39',
+'HLA-DQA1*06:02-DQB1*06:40', 'HLA-DQA1*06:02-DQB1*06:41', 'HLA-DQA1*06:02-DQB1*06:42', 'HLA-DQA1*06:02-DQB1*06:43', 'HLA-DQA1*06:02-DQB1*06:44',
+'H-2-IAb', 'H-2-IAd', 'H-2-IAk', 'H-2-IAq', 'H-2-IAs',
```
Comment (christopher-mohr) [inline epytope/EpitopePrediction/External.py]: Did you also check if it works with mouse alleles?

File: epytope/EpitopePrediction/External.py [hunk 7]
```diff
@@ -6998,7 +8327,9 @@ def parse_external_result(self, file):
                 if not row[0].isdigit():
                     continue
             
-                epitope, allele, comb_score = Peptide(row[PeptideIndex.NETCTLPAN_1_1]), Allele(row[HLAIndex.NETCTLPAN_1_1]), float(row[ScoreIndex.NETCTLPAN_1_1])
+                epitope = row[PeptideIndex.NETCTLPAN_1_1]
+                allele = row[HLAIndex.NETCTLPAN_1_1].replace('*','')
```
Comment (christopher-mohr) [inline epytope/EpitopePrediction/External.py]: A comment here would help to remember why it is necessary to change the string here.

**Review by christopher-mohr — COMMENTED** [review]
Just some comments/questions, besides that it looks good to me. Thanks 👍.
To be honest I didn't check all of the introduced available allele strings... ;)

File: epytope/EpitopePrediction/External.py (hunk 5 above)
Comment (jonasscheid) [reply epytope/EpitopePrediction/External.py]: Yes unfortunately not..only works using an actual `Allele Obj`

File: epytope/EpitopePrediction/External.py (hunk 6 above)
Comment (jonasscheid) [reply epytope/EpitopePrediction/External.py]: No I did not good point! As I checked this for all predictors I even found a bug described in #48. I'll fixed that in the next commit

Comment (jonasscheid) [reply epytope/EpitopePrediction/External.py]: No I did not good point! As I checked this for all predictors I even found a bug described in #48. I'll fixed that in the next commit

File: epytope/EpitopePrediction/External.py [hunk 8]
```diff
@@ -8432,7 +8448,7 @@ class Offset(IntEnum):
     NETMHCSTABPAN_1_0_WO_SCORE = 3
     NETMHCIIPAN_3_0 = 3
     NETMHCIIPAN_3_1 = 3
-    NETMHCIIPAN_4_0 = 3
+    NETMHCIIPAN_4_0 = 1
```
Comment (jonasscheid) [inline epytope/EpitopePrediction/External.py:8451]: As I checked every external tool again I noticed that this enum was wrong.

**Review by christopher-mohr — APPROVED** [review]
LGTM
